src/Components/Analyzers/src/ComponentParameterUsageAnalyzer.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Components.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class ComponentParameterUsageAnalyzer : DiagnosticAnalyzer
+    {
+        public ComponentParameterUsageAnalyzer()
+        {
+            SupportedDiagnostics = ImmutableArray.Create(new[]
+            {
+                DiagnosticDescriptors.ComponentParametersShouldNotBeSetOutsideOfTheirDeclaredComponent,
+            });
+        }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!ComponentSymbols.TryCreate(context.Compilation, out var symbols))
+                {
+                    // Types we need are not defined.
+                    return;
+                }
+
+                context.RegisterOperationBlockStartAction(startBlockContext =>
+                {
+                    startBlockContext.RegisterOperationAction(context =>
+                    {
+                        var leftHandSide = context.Operation.Children.FirstOrDefault();
```

Comment:
@mavasani is there a better way to get the left hand side of an assignment operation here? I tested this with several inputs and all seemed correct but want to double check.

Reply:
ISimpleAssignmentOperation, ICompoundAssignmentOperation and ICoalesceAssignmentOperation all derive from IAssignmentOperation (http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/IAssignmentOperation.cs,30dfd6516f6a7ad0,references)

I would recommend using `((IAssignmentOperation)context.Operation).Target`, see http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/IAssignmentOperation.cs,22

Reply:
Awesome, thank you @mavasani!